### PR TITLE
Add test pipeline for Mac aarch64

### LINKED ISSUE #9117: CI: add support for M1 macOS runner
### Is your feature request related to a problem or challenge?

Github Action supports M1 macOS runner, and it will be better to integrate it in the current ci. 
https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #9117.

## What changes are included in this PR?

Add new CI pipeline for Mac OSX aarch64:

Screenshot 2024-02-10 at 11 57 19 AM

### CODE REVIEW

**Comment by viirya** [comment]
https://github.com/viirya/arrow-datafusion/actions/runs/7857070647/job/21440535571

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -308,6 +308,22 @@ jobs:
           cd datafusion-cli
           cargo test --lib --tests --bins --all-features  
 
+  macos-aarch64:
+    name: cargo test (macos-aarch64)
+    runs-on: macos-14
```
Comment (alamb) [inline .github/workflows/rust.yml:313]: 👍

**Review by alamb — APPROVED** [review]
Thanks @viirya  -- this looks great to me


I took a quick look at the new CI run https://github.com/apache/arrow-datafusion/actions/runs/7857228221/job/21440868333?pr=9191

and it looks quite nice 👨‍🍳 👌

**Comment by viirya** [comment]
I verified that the bug in #9163 is caught by this new CI pipeline: https://github.com/viirya/arrow-datafusion/actions/runs/7857579868/job/21441628336

**Comment by alamb** [comment]
FYI @Jefffrey

**Comment by alamb** [comment]
Let's plan to merge this tomorrow unless others would like more time to review

**Review by Jefffrey — APPROVED** [review]
Looks good 👍

**Comment by viirya** [comment]
Thank you @alamb @Jefffrey

**Pull request merged**